### PR TITLE
verifier: readability, quality, bugfixes

### PR DESCRIPTION
**Pull request by nalinbhardwaj**
Screenshot 2023-09-08 at 4 01 45 PM

### CODE REVIEW

File: src/P256Verifier.sol [hunk 1]
```diff
@@ -58,286 +58,189 @@ contract P256Verifier {
...
-                //zz :=mulmod(zz, mload(T),p) //1/z
-                //zz:= mulmod(zz,zz,p) //1/zz
-                X := mulmod(X, mload(T), p) //X/zz
-            } //end assembly
-        } //end unchecked
-
-        return X;
+        if (scalar_u == 0 && scalar_v == 0) return (0, true);
+
+        (H0, H1, add_success) = ecAff_add(gx, gy, Q0, Q1);
```
Comment (dcposch) [inline src/P256Verifier.sol]:
```
// H = g + Q
```

File: src/P256Verifier.sol [hunk 2]
```diff
@@ -58,286 +58,189 @@ contract P256Verifier {
...
+
+        // Find the first bit index that's active in either scalar_u or scalar_v.
+        while(index >= 0) {
+            bitpair = compute_bitpair(uint256(index), scalar_u, scalar_v);
+            index--;
+            if (bitpair != 0) break;
+        }
+
+        // initialise (X, Y) depending on the first active bitpair.
+        if (bitpair == 0) { // if scalar_u = scalar_v = 0.
```
Comment (dcposch) [inline src/P256Verifier.sol]:
```
assert(bitpair > 0);
```

File: src/P256Verifier.sol [hunk 3]
```diff
@@ -58,286 +58,189 @@ contract P256Verifier {
...
+            X = 0;
+            Y = 0;
+        }
+        else if (bitpair == 1) {
+            X = gx;
+            Y = gy;
```
Comment (dcposch) [inline src/P256Verifier.sol]:
```
(X, Y) = (gx, gy);
```

File: src/P256Verifier.sol [hunk 4]
```diff
@@ -58,286 +58,189 @@ contract P256Verifier {
...
+        } else if (bitpair == 2) {
+            X = Q0;
+            Y = Q1;
```
Comment (dcposch) [inline src/P256Verifier.sol]: might be worth calling this `(QX, QY)` or conversely `g0, g1`

File: src/P256Verifier.sol [hunk 5]
```diff
@@ -58,286 +58,189 @@ contract P256Verifier {
...
+                T2 = gy;
+            } else if (bitpair == 2) {
+                T1 = Q0;
+                T2 = Q1;
+            } else {
+                T1 = H0;
+                T2 = H1;
+            }
+
+            // if (zz == 0) { // this should never happen?
```
Comment (dcposch) [inline src/P256Verifier.sol]: assert

File: src/P256Verifier.sol [hunk 6]
```diff
@@ -58,286 +58,189 @@ contract P256Verifier {
...
-        uint256 zzz0;
+        uint256 x1, // todo 1 index here
+        uint256 y1,
+        uint256 x2,
+        uint256 y2
+    ) internal view returns (uint256, uint256, bool) {
+        uint256 zz1;
+        uint256 zzz1;
 
-        if (ecAff_IsZero(x0, y0)) return (x1, y1);
```
Comment (dcposch) [inline src/P256Verifier.sol:331]: !!!

File: src/P256Verifier.sol [hunk 7]
```diff
@@ -347,107 +250,138 @@ contract P256Verifier {
...
+    function FCL_pModInv(uint256 u) internal view returns (uint256 result, bool success) {
+        return FCL_ModInv(u, p, minus_2modp);
+    }
+
+    /**
+     * @dev u^-1 mod f = u^(phi(f) - 1) mod f = u^(f-2) mod f for prime f
+     * by Fermat's little theorem, compute u^(f-2) mod f using modexp precompile
+     * Assume f != 0.
+     */
+    function FCL_ModInv(uint256 u, uint256 f, uint256 minus_2modf) internal view returns (uint256 result, bool success) {
```
Comment (dcposch) [inline src/P256Verifier.sol]: we can probably remove the FCL prefixes

**Review by dcposch — COMMENTED** [review]
Less gas
Less bugs
Less code

amazing 🚀

File: src/P256Verifier.sol (hunk 1 above)
Comment (nalinbhardwaj) [reply src/P256Verifier.sol]: done

File: src/P256Verifier.sol (hunk 3 above)
Comment (nalinbhardwaj) [reply src/P256Verifier.sol]: done

File: src/P256Verifier.sol (hunk 2 above)
Comment (nalinbhardwaj) [reply src/P256Verifier.sol]: added `//invariant` comments

File: src/P256Verifier.sol (hunk 4 above)
Comment (nalinbhardwaj) [reply src/P256Verifier.sol]: done

File: src/P256Verifier.sol (hunk 7 above)
Comment (nalinbhardwaj) [reply src/P256Verifier.sol]: done

File: src/P256Verifier.sol (hunk 5 above)
Comment (nalinbhardwaj) [reply src/P256Verifier.sol]: add invariant

**Pull request merged**